appserver/payara-appserver-modules/microprofile/fault-tolerance/src/main/java/fish/payara/microprofile/faulttolerance/service/FaultToleranceServiceImpl.java

```diff
@@ -122,20 +113,52 @@
     @Inject
     private Events events;
 
-    private final Map<String, FaultToleranceApplicationState> stateByApplication = new ConcurrentHashMap<>();
-    private ManagedScheduledExecutorService defaultScheduledExecutorService;
-    private ManagedExecutorService defaultExecutorService;
+    @Inject
+    private MetricsService metricsService;
+
+    private final ConcurrentMap<String, FaultToleranceMethodContextImpl> methodByTargetObjectAndName = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, BindableFaultToleranceConfig> configByApplication = new ConcurrentHashMap<>();
+    private ThreadPoolExecutor asyncExecutorService;
+    private ScheduledExecutorService delayExecutorService;
 
     @PostConstruct
-    public void postConstruct() throws NamingException {
+    public void postConstruct() {
         events.register(this);
-        serviceConfig = serviceLocator.getService(FaultToleranceServiceConfiguration.class);
         invocationManager = serviceLocator.getService(InvocationManager.class);
         requestTracingService = serviceLocator.getService(RequestTracingService.class);
-        InitialContext context = new InitialContext();
-        defaultExecutorService = (ManagedExecutorService) context.lookup("java:comp/DefaultManagedExecutorService");
-        defaultScheduledExecutorService = (ManagedScheduledExecutorService) context
-                .lookup("java:comp/DefaultManagedScheduledExecutorService");
+        config = serviceLocator.getService(FaultToleranceServiceConfiguration.class);
+        delayExecutorService = Executors.newScheduledThreadPool(getMaxDelayPoolSize());
+        asyncExecutorService = new ThreadPoolExecutor(0, getMaxAsyncPoolSize(), 60L, TimeUnit.SECONDS,
+                new SynchronousQueue<Runnable>(true)); // a fair queue => FIFO
+        delayExecutorService.scheduleAtFixedRate(this::cleanMethodContexts, 1L, 1L, TimeUnit.MINUTES);
```

Comment:
See note above - do we want to have some hidden configuration knobs we can turn if this rate isn't ideal?

Reply:
Good idea, I add one.